src/gpio.rs

```diff
@@ -691,10 +817,92 @@ impl<'d> Input<'d> {
         self.flex
     }
 
-    // Get the pin level.
+    /// Get the pin level.
     pub fn get_level(&self) -> Level {
         self.flex.get_level()
     }
+
+    /// Wait until the pin is high. If it is already high, return immediately.
+    #[inline]
+    pub async fn wait_for_high(&mut self) {
+        self.flex.wait_for_high().await;
+    }
+
+    /// Wait until the pin is low. If it is already low, return immediately.
+    #[inline]
+    pub async fn wait_for_low(&mut self) {
+        self.flex.wait_for_low().await;
+    }
+
+    /// Wait for the pin to undergo a transition from low to high.
+    #[inline]
+    pub async fn wait_for_rising_edge(&mut self) {
+        self.flex.wait_for_rising_edge().await;
+    }
+
+    /// Wait for the pin to undergo a transition from high to low.
+    #[inline]
+    pub async fn wait_for_falling_edge(&mut self) {
+        self.flex.wait_for_falling_edge().await;
+    }
+
+    /// Wait for the pin to undergo any transition, i.e low to high OR high to low.
+    #[inline]
+    pub async fn wait_for_any_edge(&mut self) {
+        self.flex.wait_for_any_edge().await;
+    }
+}
+
+struct InputFuture<'d> {
+    pin: Peri<'d, AnyPin>,
+}
+
+impl<'d> InputFuture<'d> {
+    fn new(pin: Peri<'d, AnyPin>, level: InterruptTrigger) -> Self {
+        // Initialize the input future with the specified pin and trigger level
+
+        // Clear any existing pending interrupt on this pin
+        pin.gpio().isfr0().write(|w| unsafe { w.bits(1 << pin.pin()) });
+        pin.gpio().icr(pin.pin()).write(|w| w.isf().isf1());
+
+        // Pin interrupt configuration
+        pin.gpio().icr(pin.pin()).modify(|_, w| match level {
+            InterruptTrigger::LevelHigh => w.irqc().irqc12(),
+            InterruptTrigger::LevelLow => w.irqc().irqc8(),
+            InterruptTrigger::RisingEdge => w.irqc().irqc9(),
+            InterruptTrigger::FallingEdge => w.irqc().irqc10(),
+            InterruptTrigger::AnyEdges => w.irqc().irqc11(),
+        });
+
+        Self { pin }
+    }
+}
+
+impl<'d> Future for InputFuture<'d> {
+    type Output = ();
+
+    fn poll(self: FuturePin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // We need to register/re-register the waker for each poll because any
+        // calls to wake will deregister the waker.
+        if self.pin.port() >= WAKERS.len() {
+            panic!("Invalid GPIO port index {}", self.pin.port());
+        }
+
+        let waker = &WAKERS[self.pin.port()];
+
+        waker.register(cx.waker());
+
+        // Double check that the pin interrupt has been disabled by IRQ handler
+        if self.pin.gpio().icr(self.pin.pin()).read().bits() & (1 << self.pin.pin()) == 0 {
+            if (INTERRUPT_DETECTED[self.pin.port()].fetch_and(!(1u32 << self.pin.pin()), Ordering::Relaxed) & (1u32 << self.pin.pin())) != 0 {
```

Comment:
Maybe pull out the mask logic so it isn't repeated?

Also we don't always need to suffix with integer types in Rust like in C, there's no implicit casting to worry about :)

```suggestion
            let mask = 1 << self.pin.pin();
            if (INTERRUPT_DETECTED[self.pin.port()].fetch_and(!mask, Ordering::Relaxed) & mask) != 0 {
```